Update Highlighting Algorithm

positiveVibes4all opened the pull request:
## Summary 
This PR updates the highlighting algorithm to not rely on setting require field match to false. This fixes instances where filters that did not originate from copy_to configurations do not highlight all fields and instead just the field that they are for. This PR also merges highlighting when there is an instance of a field to highlight and one of its related subfields as well.

Commit:
Draft Commit - Fix Highlighting

- Set require field match to true
- Add highlight queries to all relevant highlight fields per copy_to configuration
- Include subFields that are on whitelist to merge when appropriate.

Commit:
BUMP

lint-action committed:
Fix code style issues with Prettier

Commit:
Update Multi-field combine

Commit:
Merge remote-tracking branch 'origin/feature/GS-5186-Highlighting-Fix-Filters-Behavior' into feature/GS-5186-Highlighting-Fix-Filters-Behavior

Commit:
Merge branch 'main' into feature/GS-5186-Highlighting-Fix-Filters-Behavior

lint-action committed:
Fix code style issues with Prettier

Commit:
Refactor Code

- refactor highlight merge
- refactor combine multi fields function
- moved highlight merge to utils file and added  tests

lint-action committed:
Fix code style issues with Prettier

stellarhoof commented on packages/provider-elasticsearch/src/example-types/results/highlighting.test.js:
```diff
@@ -394,3 +396,55 @@ describe('containsHighlightTagRegex()', () => {
   })
 })
+
+describe('Highlight field aggregation', () => {
```
We should name the describe block after the fn is being tested. I added the article that I mostly follow when writing unit tests in notion: https://www.notion.so/Best-Practices-0d931662db624a77af8dd09cc41e01bf

stellarhoof commented on packages/provider-elasticsearch/src/example-types/results/index.js:
```diff
@@ -61,6 +61,7 @@ export default {
             nodeHighlight, // The result node's highlight configuration
             hit, // The ES result
             include: resultColumns, // The columns to return
+            node, // The subfields to return
```
Comment's wrong

stellarhoof commented on packages/provider-elasticsearch/src/schema.js:
```diff
@@ -29,17 +29,22 @@ let fromEsIndexMapping = (mapping) => {
                 dataType: type,
                 // Find the child notAnalyzedField to set up facet autocomplete vs word
                 notAnalyzedField: _.findKey({ type: 'keyword' }, fields),
+                ...(copy_to && { copy_to }),
```
No need to spread it, just add `copy_to` inline

stellarhoof commented on packages/provider-elasticsearch/src/utils/highlightUtil.js:
You probably added this util at the top because of Sam's comment on making it generic but it's really only relevant to the highlighting code so we should move it there.

stellarhoof commented on packages/provider-elasticsearch/src/utils/highlightUtil.test.js:
Can we use more compact highlight tags such as `<em></em>`? It makes it easier to follow the code.

stellarhoof commented on packages/provider-elasticsearch/src/utils/highlightUtil.test.js:
```diff
@@ -0,0 +1,58 @@
+      'foo.car.bar': ['<b class="search-highlight">foo</b> this is not merged'],
+    })
+  })
+  it('should combine subfield highlights with field highlights1', () => {
```
This function serves the same purpose as the one above. Also the field values are noisy so it's difficult to detect where the highlighted substrings are.

stellarhoof commented on packages/provider-elasticsearch/src/utils/highlightUtil.test.js:
```diff
@@ -0,0 +1,58 @@
+import { mergeHitHighlights } from './highlightUtil.js'
+
+describe('Highlight field aggregation', () => {
```
There are some important test cases we can add here:
1. Test when a highlighted string is contained in another
2. Test when a highlighted string overlaps with another

stellarhoof commented on packages/provider-elasticsearch/src/utils/highlightUtil.js:
```diff
@@ -0,0 +1,68 @@
+  F.dotEncoder.encode
+)
+
+export let mergeHitHighlights = (nodeHighlight, fields, hitHighlights) => {
```
This generic fn should only take strings only, e.g. `mergeHitHighlights(hi1, hi2, ...)`

Commit:
PR Feedback

- Updated highlight_query generation
- Updated copyTo field aggregation
- Added Tests
- Refactored other code snippets

lint-action committed:
Fix code style issues with Prettier

stellarhoof commented on packages/provider-elasticsearch/src/example-types/results/highlightUtil.js:
```diff
@@ -0,0 +1,71 @@
+let highlightWith = ({ pre_tags, post_tags }) =>
+  F.highlight(pre_tags[0], post_tags[0])
+
+let findHighlightTagsRegEx = ({ pre_tags, post_tags }) =>
```
This is the same regex as https://github.com/smartprocure/contexture/blob/main/packages/provider-elasticsearch/src/example-types/results/highlighting.js#L7

stellarhoof commented on packages/provider-elasticsearch/src/example-types/results/highlightUtil.js:
```diff
@@ -0,0 +1,71 @@
+import F from 'futil'
+
+let highlightContentRegEx = ({ pre_tags, post_tags }) =>
+  new RegExp(`(?<=${pre_tags[0]})(.*?)(?=${post_tags[0]})`, 'g')
```
Lookaheads/lookbehinds are generally expensive. You can avoid them by using `matchAll` instead of `match`. Ex:

```javascript
const regex = new RegExp('<em>(.*?)</em>', 'g')
const str = 'a <em>b</em> c <em>d e</em> f'
const matches = Array.from(str.matchAll(regex)).map(_.nth(1))
```

Also, avoid compiling (instantiating) regexes multiple times if it can be avoided. For example `new RegExp('<em>(.*?)</em>', 'g')` can be compiled once and used many times inside of a loop.

stellarhoof commented on packages/provider-elasticsearch/src/example-types/results/highlightUtil.js:
```diff
@@ -0,0 +1,71 @@
+        let highlightPostings = F.flowMap(
+          _.head,
+          getHighlightMatches({ pre_tags, post_tags }),
+          getWordListMatchesRegEx,
```
This regex shouldn't be case insensitive

stellarhoof commented on packages/provider-elasticsearch/src/example-types/results/highlightUtil.test.js:
```diff
@@ -0,0 +1,55 @@
+  post_tags: ['</em>'],
+}
+
+describe('mergeHitHighlights()', () => {
```
Another test case we should add is when there are identical substrings being highlighted more than once. Ex: `'<em>lazy</em> for was really <em>lazy</em>`

stellarhoof commented on packages/provider-elasticsearch/src/example-types/results/highlightUtil.js:
```diff
@@ -0,0 +1,71 @@
+  F.dotEncoder.encode
+)
+
+export let mergeHitHighlights = (
```
Now that I've read this fn I think it's easier if we do a single flow where we handle all the postings upfront. Ex:

```javascript
_.flow(
  _.flatMap(getHighlightedWords, getPostings),
  F.mergeRanges,
  highlightRanges,
)
```

Commit:
Update Tests & Fix Bug

- Update all tests to pass
- Fixed bug where copy to mapping was being translated when it should not have been for a reg ex pattern.
- Added unit test for bug

lint-action committed:
Fix code style issues with Prettier

Commit:
Relevant Nodes

Commit:
Merge remote-tracking branch 'origin/feature/GS-5186-Highlighting-Fix-Filters-Behavior' into feature/GS-5186-Highlighting-Fix-Filters-Behavior

lint-action committed:
Fix code style issues with Prettier
